internal/pkg/build/sources/oci_unpack_linux.go

```diff
@@ -68,5 +70,41 @@ func unpackRootfs(b *sytypes.Bundle, tmpfsRef types.ImageReference, sysCtx *type
 	os.RemoveAll(b.Rootfs())
 
 	// Unpack root filesystem
-	return umocilayer.UnpackRootfs(context.Background(), engineExt, b.Rootfs(), manifest, &mapOptions)
+	err = umocilayer.UnpackRootfs(context.Background(), engineExt, b.Rootfs(), manifest, &mapOptions)
+	if err != nil {
+		return err
+	}
+
+	// Fix permissions to allow the rootfs to be moved and deleted by the user
+	// if we performed a rootless extraction (#4517)
+	if mapOptions.Rootless {
+		sylog.Debugf("Fixing rootless permissions on %s", b.Rootfs())
+		return fixPermsRootless(b.Rootfs())
+	}
+
+	return nil
+
+}
+
+// fixPermsRootless forces permissions on the rootfs so that it can be easily
+// moved and deleted by a non-root user owner.
+func fixPermsRootless(rootfs string) (err error) {
+	err = filepath.Walk(rootfs, func(path string, f os.FileInfo, err error) error {
+		// Directories must have the owner 'rx' bits to allow traversal and reading on move, and the 'w' bit
+		// so their content can be deleted by the user when the rootfs/sandbox is deleted
+		switch mode := f.Mode(); {
+		case mode.IsDir():
+			if err := os.Chmod(path, f.Mode().Perm()|0700); err != nil {
+				return fmt.Errorf("error setting rootless sandbox permission for %s: %s", path, err)
+			}
+		case mode.IsRegular():
+			// Regular files must have the owner 'r' bit so that everything can be read in order to
```

Comment:
nitpick: what the comment says is not strictly true for moves. In order to move a file you need to be able to read and write to a directory, but not the file itself. I guess the read bit is needed for copying (the other part of the comment), and I'm wondering where exactly we need to copy these files. Is this about the case where the rootfs is created in one filesystem and later moved to a different one? That operation is in fact a copy+delete.

Reply:
Yes, I am using 'move' here to refer to either same-device move, or inter-device `copy+delete` hence needing `r`

The rootfs is initially extracted into `TMPDIR` so it'll often be an inter-device move (copy+delete).

Reply:
ah, good, just making sure I wasn't missing something there...